### DISCUSSION TITLE
Resolving dependencies from multiple merges

### DISCUSSION THREAD

**Discussion by pete0emerson** [gruntwork-io/knowledge-base#401] [Help] [answered]
We're just starting out with Terragrunt.

We merged in a depending module before the dependency had finished deploying. Now, all builds are failing because A didn't get created (not 100% sure why), but B's dependencies aren't met.

Am I right in thinking that dependencies must be live before depending modules can be merged in? How do I fix a situation where neither the depending NOR the dependency are live, thus causing all builds to fail?

I tried deleting the depending module from the S3 bucket, but that didn't seem to make a difference. I have also tried pushing directly to main instead of doing a PR, but that didn't do anything either.

Clarification note:

I don't have a circular dependency in the traditional sense. I tried to create a resource, and before it had fully come up, I merged in another resource addition, which then failed to start. A does not depend on B, simply put. A didn't exist which the CircleCI build for B was running, so B failed it's build, but was merged in. A failed to start at all, so now we're stuck.

---

<ins datetime="2022-05-02T18:16:43Z">
  <p><a href="https://gruntwork.zendesk.com/agent/tickets/108534">Tracked in ticket #108534</a></p>
</ins>

**Comment by yorinasub17** [comment, marked as answer]
The CI scripts use a diff based model to identify which modules to deploy. That is, it sees what changed on the last commit and only attempt to deploy those. This is most likely why "nothing is happening" - the commits aren't containing changes to the modules that you want the CI to deploy.

In general, you can force deployments on any module by adding a nonce changes to those modules. For example, you can generate a random hash/uuid and add that as a comment to the `terragrunt.hcl` file for the module you want to force the CI to deploy.

Given that, you can create a new commit that modifies the `terragrunt.hcl` for A with a superficial change (e.g., adding whitespace or a nonce comment) to force the CI build to attempt to deploy A. From there, you can try to see what is causing A to fail in the deployment.

Once A is deployed, you can then do the same thing to B to get it to deploy.

Note that we typically recommend using `terragrunt` locally to resolve issues like this for faster iteration. It will also allow you to perform state surgery (`state rm`, `state mv`, or `import`) to fix corruption issues as well, which I suspect would be in play here if you deleted the S3 state file. Once everything is resolved to reach steady state, then you can go back to relying on the CI builds for standard deployments.

**Reply by barcherf5** [reply]
Hi Yori! Thanks for that response. I think you missed the background part, the question of a depending module/service being deployed at the same time as the dependency. Is my understanding correct that if A doesn't exist 100%, then a build of B will fail?

**Reply by yorinasub17** [reply]
Yes that is correct. Unfortunately, the current pipeline script is not dependency aware, and thus won't be able to handle rolling out dependencies in modules simultaneously. For now, the workaround is to commit/merge the top level modules (e.g., A in your example) first, and the commit/merge the dependent modules (e.g., B).

**Reply by barcherf5** [reply]
That strikes me as odd. In the appendix of our reference architecture repo, I find this: "Similar to depends_on in Terraform, dependency blocks allow you to link two Live Terragrunt Configurations together so that Terragrunt will apply or destroy the modules in the right order when using the [run-all command](https://terragrunt.gruntwork.io/docs/features/execute-terraform-commands-on-multiple-modules-at-once/). Additionally, Terragrunt will automatically look up the outputs of the module referenced in the dependency block and make it available to use in your configuration."

And in the B module in question, I find a dependency block that references A. To reduce confusion, B is openvpn-server and A is VPC. openvpn-server explicitly depends on vpc.

Are you saying that the appendix is mistaken? If so, may I humbly suggest it be updated post-haste? :)

**Reply by yorinasub17** [reply]
The key here is "when using the `run-all` command." The pipeline scripts currently aren't set up to use `run-all` yet.

The primarily reason for this is mostly because there isn't a viable option for `run-all plan` that will make the pipeline sensible. `run-all plan` is always set up to fail when dependencies are in play (see https://github.com/gruntwork-io/terragrunt/issues/720#issuecomment-497888756 for more information on this).

Until we can identify a proper way to do `run-all plan`, we wanted to discourage usage of the pipeline to deploy multiple modules together in the same commit/merge.

That said, if you understand the limitations of `run-all plan`, then you can update the `deploy-infra.sh` script to call `run-all`. However, we haven't gone through this implementation ourselves either so we can't really provide much guidance on what that will entail.

**Reply by barcherf5** [reply]
Understood. Thank you for explaining!

**Reply by jwenz723** [reply]
Maybe the current logic in `deploy-infra.sh` could be used during the `plan` stage and then use `run-all` during the `apply` stage. This could be a little bit confusing though since the output for the `plan` and `apply` stages would be drastically different.
